File: src/main/java/com/yupi/template/service/SvgDiagramService.java
package com.yupi.template.service;

import cn.hutool.core.io.FileUtil;
import cn.hutool.core.util.StrUtil;
import com.alibaba.cloud.ai.dashscope.chat.DashScopeChatModel;
import com.yupi.template.config.SvgDiagramConfig;
import com.yupi.template.constant.PromptConstant;
import com.yupi.template.model.enums.ImageMethodEnum;
import lombok.extern.slf4j.Slf4j;
import org.springframework.ai.chat.messages.UserMessage;
import org.springframework.ai.chat.model.ChatResponse;
import org.springframework.ai.chat.prompt.Prompt;
import org.springframework.stereotype.Service;

import jakarta.annotation.Resource;
import java.io.File;

import static com.yupi.template.constant.ArticleConstant.PICSUM_URL_TEMPLATE;

/**
 * SVG 概念示意图生成服务
 * 使用 AI 生成 SVG 代码，适合概念示意、思维导图样式、关系展示等场景
 *
 * @author <a href="https://codefather.cn">编程导航学习圈</a>
 */
@Service
@Slf4j
public class SvgDiagramService implements ImageSearchService {

    @Resource
    private SvgDiagramConfig svgDiagramConfig;

    @Resource
    private DashScopeChatModel chatModel;

    @Resource
    private CosService cosService;

    @Override
    public String searchImage(String keywords) {
        // keywords 是图表需求描述
        return generateSvgDiagram(keywords);
    }

    /**
     * 生成 SVG 概念示意图
     *
     * @param requirement 示意图需求描述
     * @return 图片 URL，生成失败返回 null
     */
    public String generateSvgDiagram(String requirement) {
        if (StrUtil.isBlank(requirement)) {
            log.warn("SVG 图表需求为空");
            return null;
        }

        File tempFile = null;

        try {
            // 1. 调用 LLM 生成 SVG 代码
            String svgCode = callLlmToGenerateSvg(requirement);

            if (StrUtil.isBlank(svgCode)) {
                log.error("LLM 未生成 SVG 代码");
                return null;
            }

            // 2. 验证 SVG 格式
            if (!isValidSvg(svgCode)) {
                log.error("生成的 SVG 代码格式无效");
                return null;
            }

            // 3. 保存为临时文件
            tempFile = FileUtil.createTempFile("svg_chart_", ".svg", true);
            FileUtil.writeUtf8String(svgCode, tempFile);

            log.info("SVG 代码已保存到临时文件, size={}", tempFile.length());

            // 4. 上传到 COS
            String cosUrl = cosService.uploadFile(tempFile, svgDiagramConfig.getFolder());

            if (cosUrl != null && !cosUrl.isEmpty()) {
                log.info("SVG 概念示意图生成成功, url={}", cosUrl);
                return cosUrl;
            } else {
                log.error("上传 SVG 概念示意图到 COS 失败");
                return null;
            }

        } catch (Exception e) {
            log.error("SVG 概念示意图生成异常, requirement={}", requirement, e);
            return null;
        } finally {
            // 清理临时文件
            if (tempFile != null) {
                FileUtil.del(tempFile);
            }
        }
    }

    /**
     * 调用 LLM 生成 SVG 代码
     */
    private String callLlmToGenerateSvg(String requirement) {
        String prompt = PromptConstant.SVG_DIAGRAM_GENERATION_PROMPT
                .replace("{requirement}", requirement);

        log.info("开始调用 LLM 生成 SVG 概念示意图");

        ChatResponse response = chatModel.call(new Prompt(new UserMessage(prompt)));
        String svgCode = response.getResult().getOutput().getText().trim();

        // 提取 SVG 代码（移除可能的 markdown 代码块标记）
        svgCode = extractSvgCode(svgCode);

        return svgCode;
    }

    /**
     * 提取 SVG 代码（去除 markdown 代码块）
     */
    private String extractSvgCode(String text) {
        if (text == null) {
            return null;
        }

        // 去除 markdown 代码块标记
        text = text.replace("```xml", "").replace("```svg", "").replace("```", "").trim();

        // 确保包含 XML 声明
        if (!text.startsWith("<?xml")) {
            // 如果没有 XML 声明但有 <svg 标签，添加声明
            if (text.contains("<svg")) {
                text = "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n" + text;
            }
        }

        return text;
    }

    /**
     * 验证 SVG 格式
     */
    private boolean isValidSvg(String svgCode) {
        if (StrUtil.isBlank(svgCode)) {
            return false;
        }

        // 基本验证：包含 svg 标签
        return svgCode.contains("<svg") && svgCode.contains("</svg>");
    }

    @Override
    public ImageMethodEnum getMethod() {
        return ImageMethodEnum.SVG_DIAGRAM;
    }

    @Override
    public String getFallbackImage(int position) {
        return String.format(PICSUM_URL_TEMPLATE, position);
    }
}

File: python-backend/app/services/svg_diagram_service.py
"""SVG 概念示意图生成服务（第 5 期新增）"""

import logging
from typing import Optional
from openai import AsyncOpenAI

from app.config import settings
from app.constants.prompt import PromptConstant
from app.constants.article import ArticleConstant
from app.models.enums import ImageMethodEnum
from app.services.image_search_service import ImageSearchService
from app.schemas.image import ImageData, ImageRequest

logger = logging.getLogger(__name__)


class SvgDiagramService(ImageSearchService):
    """SVG 概念示意图生成服务"""
    
    def __init__(self):
        # 使用 DashScope
        self.client = AsyncOpenAI(
            api_key=settings.dashscope_api_key,
            base_url="https://dashscope.aliyuncs.com/compatible-mode/v1"
        )
        self.model = settings.dashscope_model
        self.default_width = settings.svg_diagram_default_width
        self.default_height = settings.svg_diagram_default_height
    
    async def search_image(self, keywords: str) -> Optional[str]:
        """此方法已废弃，请使用 get_image_data()"""
        return None
    
    async def get_image_data(self, request: ImageRequest) -> Optional[ImageData]:
        """获取图片数据"""
        requirement = request.get_effective_param(True)
        return await self.generate_svg_diagram_data(requirement)
    
    async def generate_svg_diagram_data(self, requirement: str) -> Optional[ImageData]:
        """
        生成 SVG 概念示意图数据
        
        Args:
            requirement: 示意图需求描述
            
        Returns:
            ImageData 包含 SVG 字节数据，生成失败返回 None
        """
        if not requirement or not requirement.strip():
            logger.warning("SVG 图表需求为空")
            return None
        
        try:
            # 1. 调用 LLM 生成 SVG 代码
            svg_code = await self._call_llm_to_generate_svg(requirement)
            
            if not svg_code or not svg_code.strip():
                logger.error("LLM 未生成 SVG 代码")
                return None
            
            # 2. 验证 SVG 格式
            if not self._is_valid_svg(svg_code):
                logger.error("生成的 SVG 代码格式无效")
                return None
            
            # 3. 转换为字节数据
            svg_bytes = svg_code.encode('utf-8')
            
            logger.info(f"SVG 概念示意图生成成功, size={len(svg_bytes)} bytes")
            return ImageData.from_bytes(svg_bytes, "image/svg+xml")
        except Exception as e:
            logger.error(f"SVG 概念示意图生成异常, requirement={requirement}, error={e}")
            return None
    
    async def _call_llm_to_generate_svg(self, requirement: str) -> Optional[str]:
        """调用 LLM 生成 SVG 代码"""
        prompt = PromptConstant.SVG_DIAGRAM_GENERATION_PROMPT.replace(
            "{requirement}", requirement
        )
        
        logger.info("开始调用 LLM 生成 SVG 概念示意图")
        
        response = await self.client.chat.completions.create(
            model=self.model,
            messages=[{"role": "user", "content": prompt}]
        )
        
        svg_code = response.choices[0].message.content.strip()
        
        # 提取 SVG 代码（移除可能的 markdown 代码块标记）
        svg_code = self._extract_svg_code(svg_code)
        
        return svg_code
    
    def _extract_svg_code(self, content: str) -> str:
        """提取 SVG 代码（移除 markdown 代码块标记）"""
        content = content.strip()
        
        # 移除 markdown 代码块标记
        if content.startswith("```"):
            lines = content.split("\n")
            # 跳过第一行（```svg 或 ```）
            lines = lines[1:]
            # 移除最后的 ```
            if lines and lines[-1].strip() == "```":
                lines = lines[:-1]
            content = "\n".join(lines)
        
        return content.strip()
    
    def _is_valid_svg(self, svg_code: str) -> bool:
        """验证 SVG 格式是否有效"""
        if not svg_code:
            return False
        
        svg_lower = svg_code.lower()
        
        # 检查是否包含 <svg 和 </svg> 标签
        has_svg_tag = "<svg" in svg_lower and "</svg>" in svg_lower
        
        # 检查是否包含 XML 声明（可选）
        has_xml_declaration = svg_code.strip().startswith("<?xml")
        
        # 至少要有 SVG 标签
        return has_svg_tag
    
    def get_method(self) -> ImageMethodEnum:
        """获取图片服务类型"""
        return ImageMethodEnum.SVG_DIAGRAM
    
    def get_fallback_image(self, position: int) -> str:
        """获取降级图片"""
        return ArticleConstant.PICSUM_URL_TEMPLATE.format(position)
